### PR TITLE
Gate loading datadog by forum-type

### PR DESCRIPTION
**Pull request by jimrandomh**
Datadog is a SaaS thing that collects logs of http requests via a middleware, SQL queries via an on-execute subscription in the postgres library, graphql queries, and CPU profiles. EA Forum is using it, but LW isn't. However, on LW servers, datadog still installs all its data collection hooks, and pays CPU-time costs for collecting data that it winds up discarding.

Looking at a nodejs trace of a particular logged-out front-page render, which took 833ms of total CPU time and finished after 1747ms total including database waits: I see 62ms in `Subscription._handler` (in `dd-trace`), which is serializing postgres queries for profiling, and 21ms in `_encode` (also in `dd-trace`), which is serializing graphql queries for profiling. So these two functions account for about 10% of CPU usage during SSR.

This PR gates importing `dd-trace` and other datadog-dependent libraries behind a call to `isDatadogEnabledOnSite` (which is currently a forum-gate for EA Forum). I verified with nodejs's profiler that the datadog-related CPU time is gone on an LW test server, and verified with some console-logs that the datadog-related files don't get imported. In theory, for EA Forum, datadog collection should not be affected. This needs to be tested by someone on the EA Forum side before merge, though. Some import-order things changed, in a way that might affect using datadog to measure server-startup-time in particular.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206261900995108) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by jpaddison3** [comment]
FYI @Will-Howard was OOO.

File: packages/lesswrong/lib/instanceSettings.ts [hunk 1]
```diff
@@ -189,7 +189,13 @@ export const noIndexSetting = new PublicInstanceSetting<boolean>('noindex', fals
...
 export const hasCuratedPostsSetting = new PublicInstanceSetting<boolean>("hasCuratedPosts", false, "optional");
+
 export const hasSideCommentsSetting = new PublicInstanceSetting<boolean>("comments.sideCommentsEnabled", isLWorAF, "optional");
 export const hasCommentsTableOfContentSetting = new PublicInstanceSetting<boolean>("comments.tableOfContentsEnabled", isLWorAF, "optional");
 export const hasDialoguesSetting = new PublicInstanceSetting<boolean>("dialogues.enabled", true, "optional");
 export const hasPostInlineReactionsSetting = new PublicInstanceSetting<boolean>("posts.inlineReactionsEnabled", isLWorAF, "optional");
+
+/** Whether datadog is enabled for this forum (not necessarily for the current user */
+export function isDatadogEnabledOnSite() {
+  return forumTypeSetting.get() === 'EAForum';
```
Comment (Will-Howard) [inline packages/lesswrong/lib/instanceSettings.ts:200]: Could you make this `isEAForum`? I think we're trying to remove these string checks

**Review by Will-Howard — COMMENTED** [review]
This does break most of the request tracing unfortunately. It still records database queries but is missing top level requests + graphql resolvers. The dogstatsd (custom metrics) thing is still working.

I would guess this is some kind of import order thing, which results in some packages being instrumented but not others, it could also be a more straightforward bug in how the datadogMiddleware is added.

I've given up for now but I'll try again to fix it and get back to you by Friday. You could also have a go yourself, it wouldn't be that hard to set up a trial account and [run the agent](https://docs.datadoghq.com/getting_started/agent/) on your machine, but it's probably quicker overall for me to do it seeing as it's already set up for me

**Review by Will-Howard — CHANGES_REQUESTED** [review]
I couldn't get it to work with the `require` syntax. I went through and checked out one file at a time and this is definitely the issue (e.g. if you just have the import in packages/lesswrong/server/serverStartup.ts, it works with `import` but not `require`, even with no conditional).

I would guess this is something to do with what gets included in the bundle? But I'm not  really sure. Anyway, given this, you can achieve something similar to what you want by making it conditional as to whether the tracer etc is initialised, like so:
```typescript
// eslint-disable-next-line no-restricted-imports
import tracer from "dd-trace";
// eslint-disable-next-line no-restricted-imports
import { StatsD } from "hot-shots";
import { isDatadogEnabledOnSite } from "../../lib/instanceSettings";

const datadogEnabled = isDatadogEnabledOnSite()

if (datadogEnabled) {
  tracer.init({
    hostname: process.env.IS_DOCKER ? "172.17.0.1" : undefined,
    sampleRate: 1,
    // TODO: Enabling log injection would let us associate logs with traces, but this requires setting up a json
    // logger rather than using console.log
    // logInjection: true
  });
  tracer.use('express', {
    service: 'forummagnum'
  })
}

export const dogstatsd = datadogEnabled ? new StatsD({
  host: process.env.IS_DOCKER ? "172.17.0.1" : undefined,
  prefix: 'forummagnum.'
}) : null

export const ddTracer = tracer;
```

I would suggest trying that and seeing if it kills the cpu usage as you want, and then if it doesn't we can go down a bundler rabbit hole

**Comment by Will-Howard** [comment]
I'm going to close this one in light of the [new version](https://github.com/ForumMagnum/ForumMagnum/pull/9097)

**Pull request closed without merging**